back-end/src/routes/api.js

```diff
@@ -104,8 +97,203 @@ const uploadFile = (request, response) => {
+    // For now this will be in the headers. Modify later maybe.
+    if (reqEmail) {
+      try {
+        console.debug("Querying photos");
+        
+        // Query for information based on input email. Use email index to search.
+        var queryParams = {
+          TableName: "Files",
+          IndexName: "email-file-index",
+          KeyConditionExpression: "#email = :email",
+          ExpressionAttributeNames: {
+            "#email": "email"
+          },
+          ExpressionAttributeValues: {
+            ":email": reqEmail
+          }
+        };
+        const queryData = await dynamodb.query(queryParams).promise();
+        const items = queryData.Items;
+
+        var responseData = [];
+        // Loop through photos to send all object data
+        prevTitle = "";
+        for (let i = 0; i < items.length; i++) {
+          if (items[i].title === prevTitle) {
+            continue;
+          }
+
+          // Get picture data from S3 to send in response.
+          var getParams = {
+            Bucket: BUCKET_NAME,
+            Key: items[i].file
+          };
+          var photoData = await s3.getObject(getParams).promise();
+          
+          responseObject = {
+            photo: photoData.Body.buffer,
+            title: items[i].title,
+            desc: items[i].desc,
+            likes: items[i].likes,
+            uploadDate: items[i].uploadDate
+          };
+
+          responseData.push(responseObject);
+          prevTitle = items[i].title;
+        }
+        statusCode = 200;
+        console.debug(responseData);
+        response.status(statusCode).send(responseData);
+      }
+      catch(e) {
+        console.debug("Could not retrieve photos", e);
+      }
+      console.debug("Retrieval successful");
+    }
+    else {
+      response.status(statusCode).send(message);
+    }
+  })().catch(e => {
+    console.debug("Something went wrong");
+  });
+};
+
+/**
+ * Retrieve all photos from a specific tag.
+ * @param {tag: String} request 
+ * @param {[{photo: Object, uploadDate: String, title: String}]} response 
+ */
+const getPhotoByTag = (request, response) => {
+  (async () => {
+    var statusCode = 400;
+    var message = "Empty";
+    const reqTag = request.headers.tag;
+    if (reqTag) {
+      try {
+        var queryParams = {
+          TableName: "Files",
+          IndexName: "tag-file-index",
+          KeyConditionExpression: "#tag = :tag",
+          ExpressionAttributeNames: {
+            "#tag": "tag"
+          },
+          ExpressionAttributeValues: {
+            ":tag": reqTag
+          }
+        };
+
+        const queryData = await dynamodb.query(queryParams).promise();
+        const items = queryData.Items;
+        console.debug("Data Queried");
+        console.debug(items);
 
+        var fileKeys = [];
+
+        var responseData = [];
+        // Loop through photos to send all object data
+        for (let i = 0; i < items.length; i++) {
```

Comment:
check out `array.map` [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map), for example

```js
items.map((item) => {
  // for loop body here
});
```

Reply:
ty

Reply:
Update: Believe it or not, a for loop seems easier for this situation. I'm not really mapping anything so much as adding entirely new objects into another array based on the queryData. Kept this the same for now.

Reply:
to allow array.map to work with `async`, you can do this:

```js
await Promise.all(items.map(async (item) => {
    // everything goes in here
  }));
```